### PR TITLE
feat: add byte to documents

### PR DESCRIPTION
**Pull request by samsja**
# Context

we need to be able to send documents as bytes inside of tensor (as it is wayyy smaller in bytes).

## What this pr do

- [x] add a byte field to image
- [x] refactor image loading 
- [x] add a byte field to audio
- [x] add a byte field to video
- [x] add a byte field to mesh

What need to be done:

figure it out if we should not do base64 encoding under the hood or let it to the user:

EDIT: conclusion we let the user deal with base64

### CODE REVIEW

File: docarray/typing/tensor/image/abstract_image_tensor.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from abc import ABC, abstractmethod
+
+from docarray.typing.tensor.abstract_tensor import AbstractTensor
+
+
+class AbstractImageTensor(AbstractTensor, ABC):
+    @abstractmethod
+    def bytes(self):
```
Comment (JoanFM) [inline docarray/typing/tensor/image/abstract_image_tensor.py]: why don't we add this `abstract` method directly to `AbstractTensor`? I do not see why this is related to `image` at all

**Review by JoanFM — DISMISSED** [review]

File: docarray/typing/tensor/image/abstract_image_tensor.py (hunk 1 above)
Comment (samsja) [reply docarray/typing/tensor/image/abstract_image_tensor.py]: this one is specific to Images as it converts the tensor to a PNG image and then saves it. This is different from passing the tensor to bytes. 

I don't think that all of the Tensor should have a bytes method

Comment (JohannesMessner) [reply docarray/typing/tensor/image/abstract_image_tensor.py]: > as it converts the tensor to a PNG image and then saves it
The method name does not really convey that imo

File: docarray/typing/tensor/image/abstract_image_tensor.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+from docarray.typing.tensor.abstract_tensor import AbstractTensor
+
+
+class AbstractImageTensor(AbstractTensor, ABC):
+    @abstractmethod
+    def to_bytes(self, format: str = 'PNG') -> bytes:
+        """
+        Convert image tensor to bytes.
+        """
```
Comment (JohannesMessner) [inline docarray/typing/tensor/image/abstract_image_tensor.py:12]: `format` and `return` are missing in the docstring

File: docarray/typing/tensor/image/image_ndarray.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from docarray.typing.tensor.ndarray import NdArray
+
+MAX_INT_16 = 2**15
+
+
+@_register_proto(proto_type_name='image_ndarray')
+class ImageNdArray(AbstractImageTensor, NdArray):
+    """
+    Subclass of NdArray, to represent an image tensor.
+    Adds image-specific features to the tensor.
```
Comment (JohannesMessner) [inline docarray/typing/tensor/image/image_ndarray.py:12]: what are such features?

File: docarray/typing/tensor/image/image_torch_tensor.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from docarray.typing.tensor.torch_tensor import TorchTensor, metaTorchAndNode
+
+T = TypeVar('T', bound='ImageTorchTensor')
+
+
+@_register_proto(proto_type_name='image_torch_tensor')
+class ImageTorchTensor(AbstractImageTensor, TorchTensor, metaclass=metaTorchAndNode):
+    """
+    Subclass of TorchTensor, to represent an image tensor.
+    Adds image-specific features to the tensor.
```
Comment (JohannesMessner) [inline docarray/typing/tensor/image/image_torch_tensor.py:14]: same comment about the docstring as above

File: docarray/typing/url/any_url.py [hunk 5]
```diff
@@ -103,3 +107,21 @@ def from_protobuf(cls: Type[T], pb_msg: 'str') -> T:
         :return: url
         """
         return parse_obj_as(cls, pb_msg)
+
+    def load_bytes(self, timeout: Optional[float] = None) -> bytes:
+        """Convert url to bytes. This will either load or download the file and save
+        it into a bytes object.
+        :param uri: the uri of Document
```
Comment (JohannesMessner) [inline docarray/typing/url/any_url.py:114]:
```suggestion
        :param uri: the URI of Document. Can be a local file path or a (remote) URL
```

File: docarray/typing/url/any_url.py [hunk 6]
```diff
@@ -103,3 +107,21 @@ def from_protobuf(cls: Type[T], pb_msg: 'str') -> T:
...
+        :param timeout: timeout for urlopen. Only relevant if uri is not local
```
Comment (JohannesMessner) [inline docarray/typing/url/any_url.py:115]:
```suggestion
        :param timeout: timeout for urlopen. Only relevant if URI is not local
```

**Review by JohannesMessner — CHANGES_REQUESTED** [review]

File: docarray/typing/bytes/image_bytes.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    @classmethod
+    def validate(
+        cls: Type[T],
+        value: Any,
+        field: 'ModelField',
+        config: 'BaseConfig',
+    ) -> T:
+
+        value = bytes_validator(value)
+        return cls(value)
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/typing/bytes/image_bytes.py:34]: I think if we ommit this validation then the from protobuf read will fail as it does not define the validate

File: docarray/typing/tensor/audio/abstract_audio_tensor.py [hunk 8]
```diff
@@ -1,19 +1,22 @@
...
 class AbstractAudioTensor(AbstractTensor, ABC):
-    @abstractmethod
-    def to_audio_bytes(self):
+    def to_bytes(self):
         """
         Convert audio tensor to bytes.
         """
-        ...
+        tensor = self.get_comp_backend().to_numpy(self)
+        tensor = (tensor * MAX_INT_16).astype('<h')
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/typing/tensor/audio/abstract_audio_tensor.py:18]: good point I need to add the test

File: docarray/typing/tensor/image/abstract_image_tensor.py (hunk 1 above)
Comment (samsja) [reply docarray/typing/tensor/image/abstract_image_tensor.py]: What name would you like ? Now it is `to_bytes`

File: docarray/typing/bytes/image_bytes.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        :param width: width of the image tensor.
+        :param height: height of the image tensor.
+        :param axis_layout: ordering of the different image axes.
+            'H' = height, 'W' = width, 'C' = color channel
+        :return: np.ndarray representing the image as RGB values
+        """
+
+        from PIL import Image as PILImage
+
+        raw_img = PILImage.open(BytesIO(self))
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/typing/bytes/image_bytes.py:92]: good point I need to add the test

File: docarray/typing/tensor/image/abstract_image_tensor.py (hunk 1 above)
Comment (JohannesMessner) [reply docarray/typing/tensor/image/abstract_image_tensor.py]: After checking the code i think the name is good, but it doesn't 'save' (to disk) anywhere, right?

**Review by JohannesMessner — APPROVED** [review]

File: docarray/typing/tensor/image/abstract_image_tensor.py (hunk 1 above)
Comment (samsja) [reply docarray/typing/tensor/image/abstract_image_tensor.py]: no no disk involved

File: docarray/typing/tensor/audio/abstract_audio_tensor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/typing/tensor/audio/abstract_audio_tensor.py:18]: lets keep it like this for now I believe

File: docarray/documents/audio.py [hunk 10]
```diff
@@ -86,6 +86,7 @@ class MultiModalDoc(Document):
     url: Optional[AudioUrl]
     tensor: Optional[AudioTensor]
     embedding: Optional[AnyEmbedding]
+    bytes: Optional[bytes]
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/documents/audio.py:89]: ah good point, I actually forgot to change it

**Pull request merged**